### ISSUE TITLE
Is it possible to clear average test loss?

### ISSUE THREAD

**Issue by phixerino** [davisking/dlib#1930]
I want to implement Early stopping similiar to the one in Keras, where training will be stopped after X number of epochs with no improvement in validation dataset loss error. So at the end of every epoch my code looks something like this:

```C++
trainer.test_one_step(validation_images, validation_labels);
double test_loss = trainer.get_average_test_loss();		
if (test_loss < min_test_loss)
{
	min_test_loss = test_loss;
	es_counter = 0;
}
else
	es_counter++;

if (es_counter == es_patience)
{
	cout << "Early stopping" << endl;
	....
}
```

My problem is that get_average_test_loss() returns average loss over all the test_one_step() calls, but I want only the loss value from the latest call. I thought that clear_average_loss() would fix this, but this only clears all the loss values from train_one_step() calls. Can I somehow clear average test loss or maybe do it some other way?

Thank you

**Comment by davisking** [comment]
The tooling already does this for you.  Just call `set_test_iterations_without_progress_threshold()`

Also, `get_average_test_loss()` is not the average of all the test loss values in the epoch.  It's an exponentially decayed running average over a much shorter window. 

Anyway, use `set_test_iterations_without_progress_threshold()`.  That will yield a behavior that is much better than checking an average loss value every epoch.  See http://blog.dlib.net/2018/02/automatic-learning-rate-scheduling-that.html for some more context on this.

**Comment by phixerino** [comment]
Thanks for the quick reply!

So if I don't want to shrink learning rate during training, should I use ```get_test_steps_without_progress()``` and end training when the returned value exceeds certain threshold set by me?
If yes, then I have a follow up question. When I have ```get_mini_batch_size() == 32```  and I'll call ```test_one_step(validation_images, validation_labels)```, where validation_images size is 96 and let's say there was no improvement in test loss error. Will ```get_test_steps_without_progress()``` return 1 or 3 (96/32)? I ask this because it's called test **steps** without progress, but in trainer_abstract.h it says:
> returns an estimate of how many **mini-batches** have executed without us observing a statistically significant decrease in the testing error

Also why does ```get_test_steps_without_progress()```return 0 if ```get_learning_rate_shrink_factor() == 1```?

One last thing, could you further elaborate on how ```get_average_test_loss()``` works? Because when I train my model, the train loss error stops improving after 50th epoch, but ```get_average_test_loss()``` still returns lower and lower values each epoch, even after epoch 500. It's like it calculates the average still with the big test loss errors from the first epochs.

Sorry for so many questions, but I'm really curious.

**Comment by davisking** [comment]
> So if I don't want to shrink learning rate during training, should I use get_test_steps_without_progress() and end training when the returned value exceeds certain threshold set by me?

Yes, that's what the example programs do.

One mini-batch is one step.

> Also why does get_test_steps_without_progress() return 0 if get_learning_rate_shrink_factor() == 1?

Read the docs.  It says why in there.  But this is so you can disable this stuff if you want to (which you almost certainly do not)

> One last thing, could you further elaborate on how get_average_test_loss() works? Because when I train my model, the train loss error stops improving after 50th epoch, but get_average_test_loss() still returns lower and lower values each epoch, even after epoch 500. It's like it calculates the average still with the big test loss errors from the first epochs.

That's not what `get_average_test_loss()` is doing.  It's just an exponentially decayed running average with a halflife of 200.  It is not remembering old values.
